models/nn_models.py

```diff
@@ -0,0 +1,244 @@
+        self.to(device) #TODO this doesn't make sense
+        
+        
+        # Set optimizer and scheduler
+        optimizer = optim_fn(self.parameters(), **optim_kwargs)
+        if scheduler_fn is not None: scheduler = scheduler_fn(optimizer, **scheduler_kwargs)
+
+        # Convert dataset to torch Dataset
+        train_dataset = TorchDataset(train_dataset)
+
+        # Get train and validation ids
+        n = len(train_dataset)
+        n_val = int(n * val_fraction)
+        n_train = n - n_val
+        train_ids = list(range(n_train))
+        val_ids = list(range(n_train, n))
+
+
+        # Get dataloaders
+        train_batch_size = min(batch_size, len(train_ids))
+        train_loader = torch.utils.data.DataLoader(train_dataset, 
+                                                   batch_size=train_batch_size, 
+                                                   sampler=torch.utils.data.SubsetRandomSampler(train_ids),
+                                                   collate_fn=train_dataset.collate_fn)
+        
+        if val_fraction > 0:
+            val_batch_size = min(batch_size, len(val_ids))
+            val_loader = torch.utils.data.DataLoader(train_dataset, 
+                                                    batch_size=val_batch_size,
+                                                    sampler=torch.utils.data.SubsetRandomSampler(val_ids),
+                                                    collate_fn=train_dataset.collate_fn)
+
+        # Load optimizer and scheduler
+        optimizer = optim_fn(self.parameters(), **optim_kwargs)
+        if scheduler_fn is not None: scheduler = scheduler_fn(optimizer, **scheduler_kwargs)
+
+        # Training loop
+        for epoch in range(num_epochs):
+            losses = []
+            self.train()
+            pbar = tqdm(train_loader, desc=f"Epoch {epoch+1}/{num_epochs}")
+            for batch in pbar:
+                # Set gradients to zero
+                optimizer.zero_grad()
+
+                # Send data to device
+                for key in batch: 
+                    if isinstance(batch[key], torch.Tensor):
+                        batch[key] = batch[key].to(device)
+
+                # Forward pass
+                features = {k: v for k, v in batch.items() if k != KEY_TARGET}
+                predictions = self(features).squeeze()
+                target = batch[KEY_TARGET]
+                loss = loss_fn(predictions, target, **loss_kwargs)
+
+                # Backward pass
+                loss.backward()
+                optimizer.step()
+
+                # Save loss
+                losses.append(loss.item())
+
+                pbar.set_description(
+                    f"Epoch {epoch+1}/{num_epochs} | Loss: {loss.item():.4f}"
+                )
+
+
+            if val_loader is not None and (epoch+1) % val_every_n_epochs == 0:
+                self.eval()
+
+                # Validation loop
+                with torch.no_grad():
```

Comment:
May want to set the model to "eval" here as well (this will be relevant in case you'll add dropout or batch norm later), and then after evaluation has been done set it back to "train"